chain/pruned_block_dispatcher.go

```diff
@@ -678,3 +692,42 @@ func (d *PrunedBlockDispatcher) handleResp(req, resp wire.Message,
 
 	return progress
 }
+
+// CancelRequest removes all information regarding a failed block request.
+// When for example the Peer disconnects or runs in a timeout we make sure
+// that all related information is deleted and a new request for this block
+// can be registered. Moreover will also cancel all depending goroutines.
+func (d *PrunedBlockDispatcher) CancelRequest(blockHash chainhash.Hash,
+	err error) {
+
+	d.blockMtx.Lock()
+
+	// Before removing the block hash we get the cancelChans which were
+	// registered for block requests which already had an ongoing pending
+	// request registered.
+	cancelChans, ok := d.blockQueryCancel[blockHash]
+
+	// Remove all data related to this block request to make sure the same
+	// block can be registered again in the future.
+	delete(d.blocksQueried, blockHash)
+	delete(d.blockQueryCancel, blockHash)
+
+	d.blockMtx.Unlock()
+
+	// In case there are dependant goroutines depending on this block
+	// request we make sure we cancel them.
+	if ok {
+		d.wg.Add(1)
+		go func() {
+			defer d.wg.Done()
+
+			for _, cancel := range cancelChans {
```

Comment:
I think this'll block if `cancelChans` is nil.

Reply:
Ah nvm, we check `ok` above. Alternatively, we could have the function short circuit and eliminate some of the nesting here.